pkg/kubelet/pod/mirror_manager.go

```diff
@@ -0,0 +1,373 @@
+// and deleted. These two functions should be only used in pod manager now.
+// We'll always create mirror pod when addStaticPod(), the caller should make sure that the static pod
+// is really changed. Usually kubelet config will guarantee that.
+func (m *mirrorManager) addStaticPod(staticPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.staticPods[getKey(staticPod)] = staticPod
+	glog.V(2).Infof("Static pod %q is added or updated, create new mirror pod", format.Pod(staticPod))
+	m.createMirrorPod(staticPod)
+}
+
+func (m *mirrorManager) deleteStaticPod(staticPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	key := getKey(staticPod)
+	if _, found := m.staticPods[key]; !found {
+		glog.V(4).Infof("The static pod %q is not found", format.Pod(staticPod))
+		return
+	}
+	glog.V(4).Infof("Delete static pod %q, delete the mirror pod", format.Pod(staticPod))
+	delete(m.staticPods, key)
+	// In fact using static pod directly is also ok, but it's more intuitive making sure that
+	// only mirror pod is passing into deleteMirrorPod() and finally handleMirrorPodDeletion()
+	m.deleteMirrorPod(newMirrorPod(staticPod))
+}
+
+func (m *mirrorManager) AddMirrorPod(mirrorPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	// Invalid mirror pod will be recreated in periodically cleanup if needed
+	m.mirrorPods.setMirrorPod(getKey(mirrorPod), mirrorPod)
+}
+
+func (m *mirrorManager) DeleteMirrorPod(mirrorPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	// Removed mirror pod will be recreated in periodically cleanup if needed
+	m.mirrorPods.removeMirrorPod(getKey(mirrorPod))
+}
+
+func (m *mirrorManager) GetMirrorPod(name, namespace string) *api.Pod {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.mirrorPods.getMirrorPod(getKeyByName(name, namespace))
+}
+
+func (m *mirrorManager) GetMirrorPodByUID(uid types.UID) *api.Pod {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.mirrorPods.getMirrorPodByUID(uid)
+}
+
+func (m *mirrorManager) GetMirrorOfStaticPod(staticPod *api.Pod) *api.Pod {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.mirrorPods.getMirrorPod(getKey(staticPod))
+}
+
+func (m *mirrorManager) TranslatePodUID(uid types.UID) types.UID {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	mirrorPod := m.mirrorPods.getMirrorPodByUID(uid)
+	if mirrorPod != nil {
+		if staticPod, found := m.staticPods[getKey(mirrorPod)]; found {
+			return staticPod.UID
+		}
+	}
+	return uid
+}
+
+func IsStaticPod(pod *api.Pod) bool {
+	source, err := kubetypes.GetPodSource(pod)
+	return err == nil && source != kubetypes.ApiserverSource
+}
+
+func IsMirrorPod(pod *api.Pod) bool {
+	_, ok := pod.Annotations[kubetypes.ConfigMirrorAnnotationKey]
+	return ok
+}
+
+// Create or recreate mirror pod passed in
+func (m *mirrorManager) handleMirrorPodCreation(mirrorPod *api.Pod) {
+	// Delete old mirror pod first
+	if !m.handleMirrorPodDeletion(mirrorPod) {
```

Comment:
I'm a bit concerned that this may cause the mirror pods to be recreated frequently if the watch is slow. 
 1. create a mirror pod m0
 2. sync: can't see m0 yet. delete m0 and create m1.
 3. sync: can't see m1 yet, delete m1 and create m2.
 4. ....

I think in practice, the 5s sync period should be able to accommodate the watch delay, but it's still worth keeping in mind. Instead of blindly deleting the mirror pod, maybe we can get the mirror pod from the apiserver and check if it's correct in the future.